src/main/java/net/neoforged/neoforge/client/entity/animation/json/AnimationParser.java

```diff
@@ -0,0 +1,119 @@
+    private static final Codec<AnimationTarget> TARGET_CODEC = ResourceLocation.CODEC
+            .flatXmap(
+                    name -> Optional.ofNullable(AnimationTypeManager.getTarget(name))
+                            .map(DataResult::success)
+                            .orElseGet(() -> DataResult.error(() -> String.format(
+                                    Locale.ENGLISH, "Animation target '%s' not found. Registered targets: %s",
+                                    name, AnimationTypeManager.getTargetList()))),
+                    target -> Optional.ofNullable(AnimationTypeManager.getTargetName(target))
+                            .map(DataResult::success)
+                            .orElseGet(() -> DataResult.error(() -> String.format(
+                                    Locale.ENGLISH, "Unregistered animation target '%s'. Registered targets: %s",
+                                    target, AnimationTypeManager.getTargetList()))));
+
+    private static final Codec<AnimationChannel.Interpolation> INTERPOLATION_CODEC = ResourceLocation.CODEC
+            .flatXmap(
+                    name -> Optional.ofNullable(AnimationTypeManager.getInterpolation(name))
+                            .map(DataResult::success)
+                            .orElseGet(() -> DataResult.error(() -> String.format(
+                                    Locale.ENGLISH, "Animation interpolation '%s' not found. Registered interpolations: %s",
+                                    name, AnimationTypeManager.getInterpolationList()))),
+                    target -> Optional.ofNullable(AnimationTypeManager.getInterpolationName(target))
+                            .map(DataResult::success)
+                            .orElseGet(() -> DataResult.error(() -> String.format(
+                                    Locale.ENGLISH, "Unregistered animation interpolation '%s'. Registered interpolations: %s",
+                                    target, AnimationTypeManager.getInterpolationList()))));
+
+    public static final MapCodec<AnimationChannel> CHANNEL_CODEC = TARGET_CODEC.dispatchMap(
+            "target",
+            channel -> AnimationTypeManager.getTargetFromChannelTarget(channel.target()),
+            target -> keyframeCodec(target)
+                    .listOf()
+                    .xmap(
+                            keyframes -> new AnimationChannel(target.channelTarget(), keyframes.toArray(Keyframe[]::new)),
+                            channel -> Arrays.asList(channel.keyframes()))
+                    .fieldOf("keyframes"));
+
+    private static final Codec<Pair<String, AnimationChannel>> NAMED_CHANNEL_CODEC = RecordCodecBuilder.create(
+            instance -> instance.group(
+                    Codec.STRING.fieldOf("bone").forGetter(Pair::key),
+                    CHANNEL_CODEC.forGetter(Pair::value)).apply(instance, Pair::of));
+
+    public static final Codec<AnimationDefinition> CODEC = RecordCodecBuilder.create(
```

Comment:
I understand what's going on for most of these codecs, but they should probably have some comments explaining things just for others who may have to come back to this in the future.

Reply:
What kind of explanation do you have in mind?

Reply:
Probably the same we have for more recent codec usages like biome modifiers or data attachments: how the encoded form of the object may look in JSON.